src/Servers/Kestrel/Core/src/Internal/Http/Http1ChunkedEncodingMessageBody.cs

```diff
@@ -0,0 +1,545 @@
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+using Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions.Internal;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
+{
+    /// <summary>
+    ///   http://tools.ietf.org/html/rfc2616#section-3.6.1
+    /// </summary>
+    public class Http1ChunkedEncodingMessageBody : Http1MessageBody
+    {
+        // byte consts don't have a data type annotation so we pre-cast it
+        private const byte ByteCR = (byte)'\r';
+        // "7FFFFFFF\r\n" is the largest chunk size that could be returned as an int.
+        private const int MaxChunkPrefixBytes = 10;
+
+        private long _inputLength;
+
+        private Mode _mode = Mode.Prefix;
+        private volatile bool _canceled;
+        private Task _pumpTask;
+        private Pipe _requestBodyPipe;
+        private ReadResult _readResult;
+
+        public Http1ChunkedEncodingMessageBody(bool keepAlive, Http1Connection context)
+            : base(context)
+        {
+            RequestKeepAlive = keepAlive;
+
+            _requestBodyPipe = CreateRequestBodyPipe(context);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            var dataLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, consumed).Length;
+            _requestBodyPipe.Reader.AdvanceTo(consumed, examined);
+            OnDataRead(dataLength);
+        }
+
+        public override bool TryRead(out ReadResult readResult)
+        {
+            TryStart();
+
+            var boolResult = _requestBodyPipe.Reader.TryRead(out _readResult);
+
+            readResult = _readResult;
+
+            if (_readResult.IsCompleted)
+            {
+                TryStop();
+            }
+
+            return boolResult;
+        }
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            TryStart();
+
+            try
+            {
+                var readAwaitable = _requestBodyPipe.Reader.ReadAsync(cancellationToken);
+
+                _readResult = await StartTimingReadAsync(readAwaitable, cancellationToken);
+            }
+            catch (ConnectionAbortedException ex)
+            {
+                throw new TaskCanceledException("The request was aborted", ex);
+            }
+
+            StopTimingRead(_readResult.Buffer.Length);
+
+            if (_readResult.IsCompleted)
+            {
+                TryStop();
+            }
+
+            return _readResult;
+        }
+
+        public override void Complete(Exception exception)
+        {
+            _requestBodyPipe.Reader.Complete(exception);
```

Comment:
This exception will get re-thrown for the writer, no? Why would we want that?

Reply:
Good call. I don't think we complete a PipeReader with an exception anywhere else in Kestrel. As is, PumpAsync would wind up completing the writer side with the same exception that's passed in here, but that shouldn't be important since no one is reading at that point.

Reply:
Sounds good, I'll remove the Complete here and just report the application exception.

Reply:
Misread, I should keep the call to complete, but remove the passed in exception.